File: pkg/vm-low-orbit/tests/fixtures/go/pubsub.go
//go:build pubsub

package main

//lint:file-ignore U1000 compiled file

import (
	"fmt"

	"github.com/taubyte/go-sdk/common"
	"github.com/taubyte/go-sdk/event"
	"github.com/taubyte/go-sdk/http/client"
	pubsub "github.com/taubyte/go-sdk/pubsub/node"
)

//export pubsubtest
func pubsubtest(e event.Event) uint32 {
	if e.Type() == common.EventTypePubsub {
		if err := runTestPubSub(e); err != nil {
			panic(fmt.Sprintf("runTestPubSub failed with %s", err))
		}

		return 0
	}

	h, err := e.HTTP()
	if err == nil {
		query, _ := h.Query().Get("name")
		if query == "pubstuff" {
			if err := runTestAttachPubsub(e); err != nil {
				panic(fmt.Sprintf("runTestPubSub failed with %s", err))
			}
		}
		if query == "actuallypublish" {
			if err := runTestPublish(e); err != nil {
				panic(fmt.Sprintf("runTestPublish failed with %s", err))
			}
		}
	}

	return 0
}

func runTestPubSub(e event.Event) error {
	p, err := e.PubSub()
	if err != nil {
		return err
	}

	channel, err := p.Channel()
	if err != nil {
		return err
	}

	expectedChannelName := "someChannel"
	if channel.Name() != expectedChannelName {
		return fmt.Errorf("incorrect channel got: %s expected: %s", channel.Name(), expectedChannelName)
	}

	data, err := p.Data()
	if err != nil {
		return err
	}

	if string(data) == "Hello, world" {
		c, err := client.New()
		if err != nil {
			return fmt.Errorf("create client failed with: %v", err)
		}

		req, err := c.Request("http://localhost:9090/pubsub", client.Method("POST"))
		if err != nil {
			return fmt.Errorf("create request failed with: %v", err)
		}

		resp, err := req.Do()
		if err != nil {
			return fmt.Errorf("do request failed with: %v", err)
		}
		defer resp.Body().Close()
	}

	return nil
}

func runTestPublish(e event.Event) error {
	channel, err := pubsub.Channel("someChannel")
	if err != nil {
		return err
	}

	err = channel.Publish([]byte("Hello, world"))
	if err != nil {
		return err
	}

	return nil
}

func runTestAttachPubsub(e event.Event) error {
	channel, err := pubsub.Channel("someChannel")
	if err != nil {
		return err
	}

	err = channel.Subscribe()
	if err != nil {
		return err
	}

	return nil
}

File: pkg/vm-low-orbit/tests/fixtures/rust/srcs/pubsub.rs
extern crate taubyte_sdk;

use taubyte_sdk::{
    event::{Event, EventType},
    http::Client,
    pubsub::{self, Channel},
};

#[no_mangle]
pub fn pubsubtest(event: Event) {
    match event.event_type() {
        EventType::EventTypeUndefined => {
            panic!("Event type is undefined")
        }
        EventType::EventTypeHttp => {
            let http = event.http().unwrap();
            let query = http.queries().get("name").unwrap();

            let err: Result<(), Box<dyn std::error::Error>>;
            match query.as_str() {
                "pubstuff" => {
                    err = do_attach_pubsub();
                }
                "actuallypublish" => {
                    err = do_publish();
                }
                _ => {
                    panic!("unknown query: {}", query);
                }
            }

            if err.is_err() {
                let err =
                    http.write(format!("{} failed with: {}", query, err.err().unwrap()).as_bytes());
                if err.is_err() {
                    panic!("Err writing: {}", err.err().unwrap())
                }
            }
        }
        EventType::EventTypePubsub => {
            let pubsub = event.pubsub().unwrap();
            let err = do_pubsub(pubsub);
            if err.is_err() {
                panic!("Err with do_pubsub: {}", err.err().unwrap())
            }
        }
        EventType::EventTypeP2P => {
            panic!("event type is EventTypeP2P");
        }
    }
}

fn do_pubsub(p: pubsub::Event) -> Result<(), Box<dyn std::error::Error>> {
    let channel = p.channel()?;
    assert_eq!(channel.name(), "someChannel");

    let data = p.data()?;
    if String::from_utf8(data)? == "Hello, world" {
        Client::post("http://localhost:9090/pubsub")?;
    }

    Ok(())
}

fn get_channel() -> Result<Channel, Box<dyn std::error::Error>> {
    Channel::new("someChannel".to_string())
}

fn do_publish() -> Result<(), Box<dyn std::error::Error>> {
    get_channel()?.publish("Hello, world".as_bytes())
}

fn do_attach_pubsub() -> Result<(), Box<dyn std::error::Error>> {
    get_channel()?.subscribe()
}
